uk.co.compucorp.civicrm.hrleaveandabsences/tests/phpunit/CRM/HRLeaveAndAbsences/Service/LeaveRequestTest.php

```diff
@@ -364,4 +367,126 @@ private function getDefaultParams($params = []) {
+
+    //update leave request and request a change in balance to the new balance
+    $params['id'] = $leaveRequest->id;
+    $params['change_balance'] = 1;
+
+    $leaveRequest = $this->getleaveRequestService()->create(
+      $params,
+      false
+    );
+
+    $balance = LeaveBalanceChange::getTotalBalanceChangeForLeaveRequest($leaveRequest);
+    $this->assertEquals($newBalance, $balance);
+  }
+
+  public function testBalanceChangeIsNotUpdatedForAlreadyCreatedLeaveRequestWhenChangeBalanceParameterIsFalseAndLeaveRequestDatesDidNotChange() {
+    HRJobContractFabricator::fabricate(
+      ['contact_id' => $this->leaveContact],
+      ['period_start_date' => '2016-01-01']
+    );
+
+    AbsencePeriodFabricator::fabricate([
+      'start_date' => CRM_Utils_Date::processDate('2016-01-01'),
+      'end_date'   => CRM_Utils_Date::processDate('2016-12-31'),
+    ]);
+
+    WorkPatternFabricator::fabricateWithA40HourWorkWeek(['is_default' => true]);
+
+    //Leave dates on Monday to Friday, all working days
+    $leaveDates = [
+      'from_date' => CRM_Utils_Date::processDate('2016-02-08'),
+      'to_date' => CRM_Utils_Date::processDate('2016-02-12')
+    ];
+
+    $params = $this->getDefaultParams($leaveDates);
+    $leaveRequest = LeaveRequestFabricator::fabricateWithoutValidation($params, true);
+
+    //Just to make sure that we have the expected balance change for the leave request
+    $previousBalance = LeaveBalanceChange::getTotalBalanceChangeForLeaveRequest($leaveRequest);
+    $this->assertEquals(-5, $previousBalance);
+
+    //Add a work pattern for the contact with effective date before the leave dates
+    $workPattern1 = WorkPatternFabricator::fabricateWithTwoWeeksAnd31AndHalfHours();
+    ContactWorkPatternFabricator::fabricate([
+      'contact_id' => $this->leaveContact,
+      'pattern_id' => $workPattern1->id,
+      'effective_date' => CRM_Utils_Date::processDate('2016-02-01')
+    ]);
+
+    //If the leave request is to be updated with new balance change, the balance would have changed.
+    $result = LeaveRequest::calculateBalanceChange($this->leaveContact,
+      new DateTime($params['from_date']),
+      $params['from_date_type'],
+      new DateTime($params['to_date']),
+      $params['to_date_type']
+    );
+    $newBalance = $result['amount'];
+
+    //The balance for leave request has changed since the contact work pattern has
+    //been changed for the date range that the leave was initially requested.
+    $this->assertNotEquals($previousBalance, $newBalance);
+
+    //update leave request and request that the previous balance be retained
+    //even though the balance has changed
+    $params['id'] = $leaveRequest->id;
+    $params['change_balance'] = 0;
+
+    $leaveRequest = $this->getleaveRequestService()->create(
+      $params,
+      false
+    );
+
+    $balance = LeaveBalanceChange::getTotalBalanceChangeForLeaveRequest($leaveRequest);
+    $this->assertEquals($previousBalance, $balance);
+  }
```

Comment:
What about tests for:
- When the dates changed and change_balance is 0
- When the dates changed and change_balance is 1